linkerd/cache/src/lib.rs

```diff
@@ -19,72 +19,125 @@ where
 {
     inner: S,
-    handle: Option<Handle<S, T>>,
-    idle: time::Duration,
-}
-
-#[derive(Debug)]
-struct Handle<S, T> {
-    handle: Arc<T>,
-    cache: Weak<RwLock<HashMap<T, (S, Weak<T>)>>>,
+    // Notifies entry's eviction task that a drop has occurred.
+    handle: Arc<Notify>,
 }
 
-type Services<T, S> = Arc<RwLock<HashMap<T, (S, Weak<T>)>>>;
+type Services<T, S> = RwLock<HashMap<T, (S, Weak<Notify>)>>;
 
 // === impl Cache ===
 
 impl<T, N> Cache<T, N>
 where
-    T: Eq + Hash + Send + Sync + 'static,
+    T: Clone + std::fmt::Debug + Eq + Hash + Send + Sync + 'static,
     N: NewService<T> + 'static,
     N::Service: Send + Sync + 'static,
 {
     pub fn layer(idle: time::Duration) -> impl layer::Layer<N, Service = Self> + Clone {
         layer::mk(move |inner| Self::new(idle, inner))
     }
 
-    pub fn new(idle: time::Duration, inner: N) -> Self {
-        let services = Services::default();
-
+    fn new(idle: time::Duration, inner: N) -> Self {
+        let services = Arc::new(Services::default());
         Self {
             inner,
             services,
             idle,
         }
     }
+
+    fn spawn_idle(
+        target: T,
+        idle: time::Duration,
+        cache: &Arc<Services<T, N::Service>>,
+    ) -> Arc<Notify> {
+        // Spawn a background task that holds the handle. Every time the handle
+        // is notified, it resets the idle timeout. Every time teh idle timeout
+        // expires, the handle is checked and the service is dropped if there
+        // are no active handles.
+        let handle = Arc::new(Notify::new());
+        tokio::spawn(Self::evict(
+            target,
+            idle,
+            handle.clone(),
+            Arc::downgrade(&cache),
+        ));
+        handle
+    }
+
+    #[instrument(level = "debug", skip(idle, reset, cache))]
+    async fn evict(
+        target: T,
+        idle: time::Duration,
+        mut reset: Arc<Notify>,
+        cache: Weak<Services<T, N::Service>>,
+    ) {
+        // Wait for the handle to be notified before starting to track idleness.
+        reset.notified().await;
+        debug!("Awaiting idleness");
+
+        // Wait for either the reset to be notified or the idle timeout to
+        // elapse.
+        loop {
+            futures::select_biased! {
+                // If the reset was notified, restart the timer.
+                _ = reset.notified().fuse() => {
+                    trace!("Reset");
+                }
+                _ = time::sleep(idle).fuse() => match cache.upgrade() {
+                    Some(cache) => match Arc::try_unwrap(reset) {
+                        // If this is the last reference to handle after the
+                        // idle timeout, remove the cache entry.
+                        Ok(_) => {
+                            let removed = cache.write().remove(&target).is_some();
+                            debug_assert!(removed, "Cache item must exist: {:?}", target);
```

Comment:
note that since the `JoinHandle` is dropped when spawning this task, this panic might not be propagated past the task boundary. Since it's a debug assertion, that's probably fine — it will still fail our tests as long as they're run on the single-threaded runtime. Just a note for the future.